pkg/controller/volume/persistentvolume/pv_controller_base.go

```diff
@@ -346,6 +347,62 @@ func (ctrl *PersistentVolumeController) updateClaimMigrationAnnotations(claim *v
 	return newClaim, nil
 }
 
+func (ctrl *PersistentVolumeController) updateVolumeFinalizers(volume *v1.PersistentVolume) (*v1.PersistentVolume, error) {
+	volumeClone := volume.DeepCopy()
+	modified, modifiedFinalizers := updateFinalizers(ctrl.csiMigratedPluginManager, volumeClone.Annotations, volumeClone.Finalizers)
+	if modified {
+		volumeClone.ObjectMeta.Finalizers = modifiedFinalizers
+		// Update the PV
+		newVol, err := ctrl.kubeClient.CoreV1().PersistentVolumes().Update(context.TODO(), volumeClone, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("persistent volume controller can't anneal finalizers: %v", err)
+		}
+		_, err = ctrl.storeVolumeUpdate(newVol)
+		if err != nil {
+			return nil, fmt.Errorf("persistent volume controller can't anneal finalizers: %v", err)
+		}
+		return newVol, nil
+	}
+	return volume, nil
+}
+
+func updateFinalizers(cmpm CSIMigratedPluginManager, ann map[string]string, finalizers []string) (bool, []string) {
+	if ann == nil {
+		// Volume does not have the any annotation. Ignore.
+		return false, finalizers
+	}
+	if len(finalizers) == 0 {
+		// Volume does not have any finalizers, nothing to modify
+		return false, finalizers
+	}
+	provisioner, ok := ann[pvutil.AnnDynamicallyProvisioned]
+	if !ok {
+		// Volume does not have the needed annotation. Ignore.
+		return false, finalizers
+	}
+	migratedToDriver := ann[pvutil.AnnMigratedTo]
+	// Check if migration is disabled for the plugin.
+	if !cmpm.IsMigrationEnabledForPlugin(provisioner) {
+		if migratedToDriver != "" {
+			modifiedFinalizers := make([]string, 0)
```

Comment:
These checks are already done in `updateVolumeMigrationAnnotations`. Would it be possible to put the finalizer code there? It could be as simple as:

```go
if utilfeature.DefaultFeatureGate.Enabled(features.HonorPVReclaimPolicy) {
    if slice.ContainsString(...) {
        slice.RemoveString(...)
    }
}
```

And rename it to something like `updateVolumeMigration`. This way, only one API server `Update` would be called. To be honest, it will complicate unit tests too.

Reply:
A simple RemoveString will not work since it makes a copy of the slice, instead, an in-place removal is necessary. Used the approach mentioned in https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating

Reply:
IMO premature optimization is evil and harm readability. Anyone could optimize `RemoveString` if we find out it's taking too much time. Please use `RemoveString`, your code calls it anyway.

Reply:
Sure, I needed to change the parameter to accept the pointer to the finalizer, fixed it in the latest change.